### ISSUE TITLE
Any plan to provide a python interface?

### ISSUE THREAD

**Issue by zmonoid** [NVIDIA/nvcomp#9]
As titled.

**Comment by nv-dlasalle** [comment]
This is not currently planned for our next release, but is something that is on our radar. What kind of objects are you thinking of compressing? Numpy arrays, tensors from a DL framework, etc.?

**Comment by zmonoid** [comment]
@nv-dlasalle, tensor from DL framework.

Currently, most of large dataset in deep learning is in image format. The bottlenet of training speed for some algorithm is actually the dataloading time. If we can compress the image data and store them in GPU memory, and decompress image data during training, some algorithm's speed will be fly high.

**Comment by nsakharnykh** [comment]
+@kkraus14

**Comment by xavirostudio** [comment]
> @nv-dlasalle, tensor from DL framework.
> 
> Currently, most of large dataset in deep learning is in image format. The bottlenet of training speed for some algorithm is actually the dataloading (host->gpu) time. If we can compress the image data and store them in GPU memory, and decompress image data (in GPU memory) during training, some algorithm's speed will be fly high.

Most of the images for training aren't stored as bitmap (RAW), the bottleneck is actually due to the fact that the images are decompressed into RGB on CPU and converted to tensors before being sent to GPU. In theory it should be possible to load the image file onto GPU and decode it using GPU.

Check nVidia's DALI - https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/image_processing/decoder_examples.html#Image-Decoder-(Hybrid) 

So you won't have wait for a python binding for this lib. ;-)

**Comment by zmonoid** [comment]
@xavirostudio Maybe I have not made my self clear.
I mean, the whole image dataset (in comopressed form) is preloaded onto GPU memory before training (not one batch). Therefore, during training, there is no host to gpu time.

**Comment by xavirostudio** [comment]
> @xavirostudio Maybe I have not made myself clear.
> I mean, the whole image dataset (in comopressed format) is preloaded onto GPU memory before training (not one batch). Therefore, during training, there is no host to gpu time.
> 
> The hybrid image decoder you mentioned in DALI works faster than CPU decoder because:
> In CPU mode, per batch data transferred from host to GPU memory is in decompressed format.
> While in hybrid mode, it is in compressed format, which is much smaller in size than in CPU mode.
> 
> If the whole compressed image dataset is preloaded onto GPU memory before training, it will be even faster since it has zero host to GPU time.

ImageNet has over 1 mil images and it's tgz archive is ~150GB - so it would obviously not fit in any GPU memory any time soon. 

Even using a smaller dataset, it would be a waste of GPU memory to load it all in GPU. The more GPU memory you have the bigger the batch size could be or the more parameters the "model" could have. 

Keep in mind that during training the GPU memory is heavily used for storing the gradients.

A way of estimating the amount of memory a batch would use is something like this: 

Batch Size x Parameters x (16 for FP64, 8 for FP32, 4 for FP16)

Let's say you have a batch size of 32 and you want to train ResNet50 (25,636,712 params) using FP32. 

It would use about 6.5 GB of GPU memory just for the parameters - in reality it would use about 7.5 GB because you also hold the RGB tensors, etc etc ... 

Most of DataLoaders know how to pre-load batches in such way that the GPU itself would be utilized as much as possible. (Sometimes a bigger batch-size doesn't mean faster convergence - but that's a different topic)

When does GPU compression comes handy then? Well, a bottleneck of data-transfer is when you train a network on multiple GPUs (or on a cluster), and the parameters have to be shared and updated between the GPUs. If you'd manage to somehow get a compression ratio of good enough to result in something like Time to compress + Time to transfer compressed + Time to decompress << Time to transfer uncompressed, you are going to have faster training.

There are other applicability of compression but in the context of training neural-networks, data-transfer between training nodes would be a good candidate.

**Comment by zmonoid** [comment]
@xavirostudio Thanks for you explaination. 

In deep reinforcement learning scenario, both network size and batch size is quite small, so GPU memory usage is quite small, and the bottlenet is at host to gpu memory time from my analysis. If there exists GPU data compression/decompression operator, the whole replay memory be stored in GPU, DRL algorithm's training speed will be much higher.

I believe there are other scenarios where you need to have a small network and batch size.

**Comment by xavirostudio** [comment]
@zmonoid sorry but I don't have much expertise in RL. I thought that most RL engines already use GPU to generate and store the environments - hence there wouldn't be a data transfer to GPU from host. If that's not the case I would look into having the "Environment" engine ported to GPU ;-) 

On other cases of small network / batch-size the "data loader" should compensate by pre-loading the batches into GPU memory. 
Depending on the framework you use to load the data you should check the pre-loading options. For example in PyTorch you'd adjust the prefetch_factor of the DataLoader or for Tensorflow you'd use tf.data.Dataset.prefetch ...

Compression would work as long as the formula is valid - Time to compress + Time to transfer compressed + Time to decompress should be lower than simply the time it would take to Transfer uncompressed data...

**Comment by technillogue** [comment]
:+1:, I have scale-to-zero inference and this could speed up cold starts significantly since they require moving large models
